src/extensions/anyone_can_pay/mod.rs

```diff
@@ -0,0 +1,236 @@
+                        let data = tx.outputs_data().get(cell_index as usize).unwrap();
+
+                        DetailedLiveCell {
+                            block_number: block.number(),
+                            block_hash: block.hash(),
+                            tx_index: tx_index as u32,
+                            cell_output: cell,
+                            cell_data: data,
+                        }
+                    } else {
+                        self.get_live_cell_by_out_point(&out_point)?
+                    };
+
+                    if self.is_acp_cell(&cell.cell_output) {
+                        let key = self.get_acp_pubkey_hash(&cell.cell_output.lock().args());
+                        acp_map
+                            .0
+                            .entry(key)
+                            .or_insert_with(Default::default)
+                            .push_removed(&out_point);
+                    }
+                }
+            }
+
+            for (idx, output) in tx.outputs().into_iter().enumerate() {
+                if self.is_acp_cell(&output) {
+                    let key = self.get_acp_pubkey_hash(&output.lock().args());
+                    let out_point = packed::OutPointBuilder::default()
+                        .tx_hash(tx.hash())
+                        .index((idx as u32).pack())
+                        .build();
+
+                    acp_map
+                        .0
+                        .entry(key)
+                        .or_insert_with(Default::default)
+                        .push_added(out_point);
+                }
+            }
+        }
+
+        self.store_acp_cells(acp_map, block.number(), &block.hash(), false)?;
+
+        Ok(())
+    }
+
+    fn rollback(&self, tip_number: BlockNumber, tip_hash: &packed::Byte32) -> Result<()> {
+        let block_key = Key::Block(tip_number, tip_hash).into_vec();
+        let raw_data = self
+            .store
+            .get(&block_key)?
+            .expect("ACP extension rollback data is not exist");
+
+        let acp_map = rlp::decode::<ACPMap>(&raw_data).unwrap();
+        self.store_acp_cells(acp_map, tip_number, &tip_hash, true)?;
+        Ok(())
+    }
+
+    fn prune(
+        &self,
+        tip_number: BlockNumber,
+        _tip_hash: &packed::Byte32,
+        keep_num: u64,
+    ) -> Result<()> {
+        let prune_to_block = tip_number - keep_num;
+        let block_key_prefix = vec![KeyPrefix::Block as u8];
+        let mut batch = self.get_batch()?;
+
+        let block_iter = self
+            .store
+            .iter(&block_key_prefix, IteratorDirection::Forward)?
+            .filter(|(key, _v)| {
+                key.starts_with(&block_key_prefix)
+                    && BlockNumber::from_be_bytes(to_fixed_array(&key[1..9])) < prune_to_block
+            });
+
+        for (key, _val) in block_iter {
+            batch.delete(key)?;
+        }
+
+        batch.commit()?;
+
+        Ok(())
```

Comment:
I don't understand why the acp cells don't rollback when pruning

Reply:
Pruning will only remove the outdated data that prepared for rollback.

Reply:
Well, I understand it now